site-src/geps/gep-726.md

```diff
@@ -0,0 +1,319 @@
+# GEP-726: Add Path Redirects and Rewrites
+
+* Issue: [#726](https://github.com/kubernetes-sigs/gateway-api/issues/726)
+* Status: Implementable
+
+## TLDR
+
+This GEP proposes adding support for path redirects and rewrites in addition to
+host rewrites. This would augment the existing host redirection capabilities.
+
+## Goals
+
+* Implement path redirects.
+* Implement the most portable and simple forms of path rewrites.
+* Describe how more advanced rewrite and redirect and redirect capabilities
+  could be added in the future.
+
+## API
+
+Although many implementations support very advanced rewrite and redirect
+capabilities, the following are the most [portable](#portability) concepts that
+are not already supported by the Gateway API:
+
+* Path redirects
+* Path prefix redirects
+* Path prefix rewrites
+* Host rewrites
+
+Although regular expression based redirects and rewrites are commonly supported,
+there is significantly more variation in both if and how they are implemented.
+Given the wide support for this concept, it is important to design the API in a
+way that would make it easy to add this capability in the future.
+
+### Path Modifiers
+
+Both redirects and rewrites would share the same `PathModifier` types:
+
+```go
+// HTTPPathModifierType defines the type of path redirect.
+type HTTPPathModifierType string
+
+const (
+  // This type of modifier indicates that the complete path will be replaced by
+  // the path redirect value.
+  AbsoluteHTTPPathModifier HTTPPathModifierType = "Absolute"
+
+  // This type of modifier indicates that any prefix path matches will be
+  // replaced by the substitution value. For example, a path with a prefix match
+  // of "/foo" and a ReplacePrefixPathMatch substitution of "/bar" will have the
```

Comment:
From an API perspective, I think that I would prefer to decouple prefix matching from prefix rewriting. This is a  bit awkward in Envoy IIRC, but it's doable (at least Contour does it).

If we keep this as it is, then a rewrite would become invalid if the path match changes, and it would also rewrite differently depending on the match prefix. This seems operationally fraught to me, and likely to trick people with unexpected side-effects.

Reply:
My understanding of the [Envoy spec](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-redirectaction-prefix-rewrite) is that the only kind of prefix rewriting it supports is based on the corresponding prefix match. That's also true for the `pathPrefixRewrite` in [GCP UrlMaps](https://cloud.google.com/compute/docs/reference/rest/v1/urlMaps#UrlMap.FIELDS.inlinedField_22). Of course with Envoy and other proxies, regex path rewrites could be used to enable more advanced use cases including rewrites based on an arbitrary prefix (covered later in this GEP as possible future extensions), but that doesn't seem to be quite as widely supported as rewriting based on a prefix match.

Reply:
Contour implements this by implicitly expanding it to multiple matches (IIRC I wrote that, but I forget the details). As an API feature, I'd prefer that Envoy limitations didn't leak through, though I'm sympathetic to the complexity of implementing this in the Envoy API.

Reply:
>I'd prefer that Envoy limitations didn't leak through

I agree with that, though in this case it's something that also affects GCP APIs, where the proposed `PrefixMatch` is the only path rewrite currently supported. Although I don't want to copy APIs, I also want to build an API that is broadly implementable. I think the current GEP along with the proposed future extensions for Prefix and Regex rewrites based on custom patterns will provide the best possible balance between an implementable and understandable API, but open to other potential approaches.